File: evolution-experiment/src/lib/simulation/creature.ts
import { v4 as uuidv4 } from 'uuid';
import type { Point2, Vector2 } from './math';
import { vecSub, vecAdd, vecScale, vecNorm, vecNormalize, vecNeg, distanceToLine } from './math';
import type { RNG } from './rng';
import type { Edible, FoodType } from './food';

const ENERGY_COST_SCALE_FACTOR = 1 / 10000;

export enum ObjectiveIntensity {
  MinorCraving = 0,
  MinorAversion = 1,
  ModerateCraving = 2,
  ModerateAversion = 3,
  MajorCraving = 4,
  MajorAversion = 5,
  VitalCraving = 6,
  VitalAversion = 7,
}

export function isAversion(intensity: ObjectiveIntensity): boolean {
  return (
    intensity === ObjectiveIntensity.MinorAversion ||
    intensity === ObjectiveIntensity.ModerateAversion ||
    intensity === ObjectiveIntensity.MajorAversion ||
    intensity === ObjectiveIntensity.VitalAversion
  );
}

export interface Objective {
  pos: Point2;
  intensity: ObjectiveIntensity;
  reason: string;
}

export interface MutatableTrait {
  value: number;
  variance: number;
}

export type CreatureState = 'active' | 'asleep' | 'dead';

export interface FoodRecord {
  step: number;
  id: string;
  type: FoodType;
}

export interface Creature {
  id: string;
  speed: MutatableTrait;
  size: MutatableTrait;
  senseRange: MutatableTrait;
  reach: MutatableTrait;
  fleeDistance: MutatableTrait;
  lifeSpan: MutatableTrait;

  energy: number;
  energyConsumed: number;
  age: number;
  pos: Point2;
  homePos: Point2;
  foodsEaten: FoodRecord[];
  movementHistory: Point2[];
  state: CreatureState;
  objective: Objective | null;
}

function mutatePositiveNonZero(rng: RNG, center: number, variance: number): number {
  return Math.max(Number.MIN_VALUE, rng.gaussian(center, variance));
}

function mutatePositive(rng: RNG, center: number, variance: number): number {
  return Math.max(0, rng.gaussian(center, variance));
}

export function createDefaultCreature(pos: Point2): Creature {
  return {
    id: uuidv4(),
    speed: { value: 1.0, variance: 1.0 },
    size: { value: 1.0, variance: 1.0 },
    senseRange: { value: 1.0, variance: 1.0 },
    reach: { value: 1.0, variance: 1.0 },
    fleeDistance: { value: 1.0, variance: 1.0 },
    lifeSpan: { value: 1.0, variance: 1.0 },
    energy: 500.0,
    energyConsumed: 0.0,
    age: 0,
    pos: [pos[0], pos[1]],
    homePos: [pos[0], pos[1]],
    foodsEaten: [],
    movementHistory: [[pos[0], pos[1]]],
    state: 'active',
    objective: null,
  };
}

export function createCreatureWithTraits(
  pos: Point2,
  speed: MutatableTrait,
  size: MutatableTrait,
  senseRange: MutatableTrait,
  reach: MutatableTrait,
  fleeDistance: MutatableTrait,
  lifeSpan: MutatableTrait,
  energy: number = 500.0,
): Creature {
  return {
    id: uuidv4(),
    speed,
    size,
    senseRange,
    reach,
    fleeDistance,
    lifeSpan,
    energy,
    energyConsumed: 0.0,
    age: 0,
    pos: [pos[0], pos[1]],
    homePos: [pos[0], pos[1]],
    foodsEaten: [],
    movementHistory: [[pos[0], pos[1]]],
    state: 'active',
    objective: null,
  };
}

export function withNewPosition(c: Creature, pos: Point2): Creature {
  return {
    ...c,
    id: uuidv4(),
    pos: [pos[0], pos[1]],
    homePos: [pos[0], pos[1]],
    movementHistory: [[pos[0], pos[1]]],
    foodsEaten: [],
    energyConsumed: 0.0,
    age: c.age,
    state: 'active',
    objective: null,
  };
}

export function mutateCreature(c: Creature, rng: RNG): Creature {
  const homePos: Point2 = [c.homePos[0], c.homePos[1]];
  return {
    ...createDefaultCreature(homePos),
    speed: { value: mutatePositiveNonZero(rng, c.speed.value, c.speed.variance), variance: c.speed.variance },
    size: { value: mutatePositiveNonZero(rng, c.size.value, c.size.variance), variance: c.size.variance },
    senseRange: { value: mutatePositive(rng, c.senseRange.value, c.senseRange.variance), variance: c.senseRange.variance },
    reach: { value: mutatePositiveNonZero(rng, c.reach.value, c.reach.variance), variance: c.reach.variance },
    fleeDistance: { value: mutatePositive(rng, c.fleeDistance.value, c.fleeDistance.variance), variance: c.fleeDistance.variance },
    lifeSpan: { value: mutatePositiveNonZero(rng, c.lifeSpan.value, c.lifeSpan.variance), variance: c.lifeSpan.variance },
    energy: c.energy,
  };
}

export function growOlder(c: Creature): Creature {
  const homePos: Point2 = [c.homePos[0], c.homePos[1]];
  return {
    ...createDefaultCreature(homePos),
    id: c.id,
    speed: { ...c.speed },
    size: { ...c.size },
    senseRange: { ...c.senseRange },
    reach: { ...c.reach },
    fleeDistance: { ...c.fleeDistance },
    lifeSpan: { ...c.lifeSpan },
    energy: c.energy,
    age: c.age + 1,
  };
}

export function getSize(c: Creature): number {
  return c.size.value;
}

export function getSpeed(c: Creature): number {
  return c.speed.value * c.size.value / 10;
}

export function getSenseRange(c: Creature): number {
  return c.senseRange.value;
}

export function getReach(c: Creature): number {
  return Math.max(c.reach.value, getSize(c) / 4);
}

export function getLifeSpan(c: Creature): number {
  return c.lifeSpan.value;
}

export function getMotionEnergyCost(c: Creature): number {
  const s = getSize(c);
  const sp = getSpeed(c);
  const sr = getSenseRange(c);
  return ENERGY_COST_SCALE_FACTOR * (s * s * s * sp * sp + sr);
}

export function getEnergyLeft(c: Creature): number {
  return Math.max(0, c.energy - c.energyConsumed);
}

export function applyEnergyCost(c: Creature, cost: number): void {
  c.energyConsumed += cost;
  if (getEnergyLeft(c) <= 0) {
    c.state = 'dead';
  }
}

export function moveTo(c: Creature, pos: Point2): void {
  c.pos = [pos[0], pos[1]];
  c.movementHistory.push([pos[0], pos[1]]);
  const cost = getMotionEnergyCost(c);
  applyEnergyCost(c, cost);
}

export function getLastPosition(c: Creature): Point2 | null {
  const len = c.movementHistory.length;
  if (len <= 1) return null;
  return c.movementHistory[len - 2];
}

export function getDirection(c: Creature): Vector2 {
  let disp: Vector2 | null = null;

  if (c.objective) {
    let d = vecSub(c.objective.pos, c.pos);
    if (isAversion(c.objective.intensity)) {
      d = vecScale(d, -1);
    }
    if (vecNorm(d) !== 0) {
      disp = d;
    }
  }

  if (!disp) {
    const last = getLastPosition(c);
    if (last) {
      disp = vecSub(c.pos, last);
    }
    if (!disp || vecNorm(disp) === 0) {
      disp = [1, 0];
    }
  }

  return vecNormalize(disp);
}

export function canSee(c: Creature, pt: Point2): boolean {
  return vecNorm(vecSub(pt, c.pos)) <= getSenseRange(c);
}

export function canReachNow(c: Creature, pt: Point2): boolean {
  return vecNorm(vecSub(pt, c.pos)) <= getReach(c);
}

export function canReach(c: Creature, pt: Point2): boolean {
  if (canReachNow(c, pt)) return true;

  const last = getLastPosition(c);
  if (!last) return false;

  const d = distanceToLine(c.pos, last, pt);
  return d !== null && d <= getReach(c);
}

export function withinFleeDistance(c: Creature, pt: Point2): boolean {
  if (canSee(c, pt)) {
    const d = vecNorm(vecSub(c.pos, pt));
    return d < c.fleeDistance.value;
  }
  return false;
}

export function addObjective(c: Creature, obj: Objective): void {
  if (!c.objective) {
    c.objective = obj;
    return;
  }
  if (obj.intensity > c.objective.intensity) {
    c.objective = obj;
    return;
  }
  if (obj.intensity === c.objective.intensity) {
    const distOld = vecNorm(vecSub(c.pos, c.objective.pos));
    const distNew = vecNorm(vecSub(c.pos, obj.pos));
    if (distNew < distOld) {
      c.objective = obj;
    }
  }
}

export function resetObjective(c: Creature): void {
  c.objective = null;
}

export function eatFood(c: Creature, step: number, food: Edible): void {
  c.foodsEaten.push({ step, id: food.getEdibleId(), type: food.getType() });
}

export function sleep(c: Creature): void {
  c.state = 'asleep';
}

export function kill(c: Creature): void {
  c.state = 'dead';
}

export function isAlive(c: Creature): boolean {
  return c.state !== 'dead';
}

export function isActive(c: Creature): boolean {
  return c.state === 'active';
}

File: evolution_experiment/simulator/creature.py
"""
Creature class: traits, mutation, energy, movement, objectives.
Ported from src/wasm/src/creature/mod.rs and creature/mutatable.rs.
"""

import sys
import math
import numpy as np
from enum import IntEnum
from .math_utils import distance_to_line

ENERGY_COST_SCALE_FACTOR = 1.0 / 10_000.0
FLOAT_MIN_POSITIVE = sys.float_info.min


class ObjectiveIntensity(IntEnum):
    """Ordered lowest to highest, matching the Rust enum derive(PartialOrd)."""
    MinorCraving = 0
    MinorAversion = 1
    ModerateCraving = 2
    ModerateAversion = 3
    MajorCraving = 4
    MajorAversion = 5
    VitalCraving = 6
    VitalAversion = 7


_AVERSION_SET = frozenset({
    ObjectiveIntensity.MinorAversion,
    ObjectiveIntensity.ModerateAversion,
    ObjectiveIntensity.MajorAversion,
    ObjectiveIntensity.VitalAversion,
})


class CreatureState(IntEnum):
    DEAD = 0
    ASLEEP = 1
    ACTIVE = 2


class Objective:
    __slots__ = ('pos', 'intensity', 'reason')

    def __init__(self, pos, intensity, reason):
        self.pos = np.asarray(pos, dtype=np.float64)
        self.intensity = intensity
        self.reason = reason


class Food:
    __slots__ = ('position', 'eaten', 'eaten_step')

    def __init__(self, position):
        self.position = np.asarray(position, dtype=np.float64)
        self.eaten = False
        self.eaten_step = None

    def is_eaten(self):
        return self.eaten

    def mark_eaten(self, step):
        self.eaten = True
        self.eaten_step = step


class Creature:
    """Faithful port of the Rust Creature struct.

    Traits stored as (value, variance) tuples for mutation.
    Default values from src/store/simulation.js:
        speed: (10, 0.5), size: (10, 0.5), sense_range: (20, 0.5),
        reach: (1, 0), flee_distance: (1e12, 0), life_span: (1e4, 0),
        energy: 500
    """

    __slots__ = (
        'pos', 'home_pos', 'speed', 'size', 'sense_range_trait',
        'reach_trait', 'flee_distance', 'life_span', 'energy',
        'energy_consumed', 'age', 'foods_eaten', '_prev_pos',
        'state', 'objective',
        '_eff_speed', '_eff_reach', '_eff_size', '_eff_sense',
        '_energy_cost',
    )

    def __init__(self, pos, speed=(10.0, 0.5), size=(10.0, 0.5),
                 sense_range=(20.0, 0.5), reach=(1.0, 0.0),
                 flee_distance=(1e12, 0.0), life_span=(1e4, 0.0),
                 energy=500.0, age=0):
        self.pos = np.asarray(pos, dtype=np.float64).copy()
        self.home_pos = self.pos.copy()
        self.speed = (float(speed[0]), float(speed[1]))
        self.size = (float(size[0]), float(size[1]))
        self.sense_range_trait = (float(sense_range[0]), float(sense_range[1]))
        self.reach_trait = (float(reach[0]), float(reach[1]))
        self.flee_distance = (float(flee_distance[0]), float(flee_distance[1]))
        self.life_span = (float(life_span[0]), float(life_span[1]))
        self.energy = float(energy)
        self.energy_consumed = 0.0
        self.age = int(age)
        self.foods_eaten = []
        self._prev_pos = None
        self.state = CreatureState.ACTIVE
        self.objective = None
        self._cache_traits()

    def _cache_traits(self):
        self._eff_size = self.size[0]
        self._eff_speed = self.speed[0] * self._eff_size / 10.0
        self._eff_sense = self.sense_range_trait[0]
        self._eff_reach = max(self.reach_trait[0], self._eff_size / 4.0)
        self._energy_cost = ENERGY_COST_SCALE_FACTOR * (
            self._eff_size ** 3 * self._eff_speed ** 2 + self._eff_sense)

    # --- Effective trait accessors (match Rust getters) ---

    def get_speed(self):
        return self._eff_speed

    def get_size(self):
        return self._eff_size

    def get_sense_range(self):
        return self._eff_sense

    def get_reach(self):
        return self._eff_reach

    def get_life_span(self):
        return self.life_span[0]

    # --- Energy ---

    def get_motion_energy_cost(self):
        return self._energy_cost

    def get_energy_left(self):
        return max(self.energy - self.energy_consumed, 0.0)

    def apply_energy_cost(self, cost):
        self.energy_consumed += cost
        if self.get_energy_left() <= 0.0:
            self.state = CreatureState.DEAD

    # --- State queries ---

    def is_alive(self):
        return self.state != CreatureState.DEAD

    def is_active(self):
        return self.state == CreatureState.ACTIVE

    # --- Movement ---

    def move_to(self, pos):
        self._prev_pos = self.pos.copy()
        self.pos = np.asarray(pos, dtype=np.float64).copy()
        self.apply_energy_cost(self.get_motion_energy_cost())

    def get_last_position(self):
        return self._prev_pos

    def get_direction(self):
        if self.objective is not None:
            dx = float(self.objective.pos[0]) - float(self.pos[0])
            dy = float(self.objective.pos[1]) - float(self.pos[1])
            if self.objective.intensity in _AVERSION_SET:
                dx, dy = -dx, -dy
            n = math.sqrt(dx * dx + dy * dy)
            if n != 0.0:
                return np.array([dx / n, dy / n])
        last = self._prev_pos
        if last is not None:
            dx = float(self.pos[0]) - float(last[0])
            dy = float(self.pos[1]) - float(last[1])
            n = math.sqrt(dx * dx + dy * dy)
            if n != 0.0:
                return np.array([dx / n, dy / n])
        return np.array([1.0, 0.0])

    # --- Objectives ---

    def add_objective(self, obj):
        if self.objective is None:
            self.objective = obj
        elif obj.intensity > self.objective.intensity:
            self.objective = obj
        elif obj.intensity == self.objective.intensity:
            ox = float(self.pos[0]) - float(self.objective.pos[0])
            oy = float(self.pos[1]) - float(self.objective.pos[1])
            nx = float(self.pos[0]) - float(obj.pos[0])
            ny = float(self.pos[1]) - float(obj.pos[1])
            if (nx * nx + ny * ny) < (ox * ox + oy * oy):
                self.objective = obj

    def reset_objective(self):
        self.objective = None

    # --- Sensing ---

    def can_see(self, pt):
        dx = float(pt[0]) - float(self.pos[0])
        dy = float(pt[1]) - float(self.pos[1])
        return (dx * dx + dy * dy) <= self.get_sense_range() ** 2

    def can_reach_now(self, pt):
        dx = float(pt[0]) - float(self.pos[0])
        dy = float(pt[1]) - float(self.pos[1])
        return (dx * dx + dy * dy) <= self.get_reach() ** 2

    def can_reach(self, pt):
        reach = self.get_reach()
        dx = float(pt[0]) - float(self.pos[0])
        dy = float(pt[1]) - float(self.pos[1])
        if math.sqrt(dx * dx + dy * dy) <= reach:
            return True
        last = self._prev_pos
        if last is None:
            return False
        r1x = float(self.pos[0])
        r1y = float(self.pos[1])
        r2x = float(last[0])
        r2y = float(last[1])
        px = float(pt[0])
        py = float(pt[1])
        vx = r2x - r1x
        vy = r2y - r1y
        v_norm = math.sqrt(vx * vx + vy * vy)
        if v_norm == 0.0:
            return False
        nx = vx / v_norm
        ny = vy / v_norm
        pax = r1x - px
        pay = r1y - py
        pbx = r2x - px
        pby = r2y - py
        pa_dot_n = pax * nx + pay * ny
        pb_dot_n = pbx * nx + pby * ny
        if pa_dot_n * pb_dot_n > 0.0:
            return False
        proj_x = -pa_dot_n * nx
        proj_y = -pa_dot_n * ny
        diff_x = proj_x - (px - r1x)
        diff_y = proj_y - (py - r1y)
        return math.sqrt(diff_x * diff_x + diff_y * diff_y) <= reach

    def within_flee_distance(self, pt):
        dx = float(pt[0]) - float(self.pos[0])
        dy = float(pt[1]) - float(self.pos[1])
        d_sq = dx * dx + dy * dy
        if d_sq > self.get_sense_range() ** 2:
            return False
        return d_sq < self.flee_distance[0] ** 2

    # --- Actions ---

    def eat_food(self, step, food_type="food"):
        self.foods_eaten.append((step, food_type))

    def sleep(self):
        self.state = CreatureState.ASLEEP

    def kill(self):
        self.state = CreatureState.DEAD

    # --- Reproduction ---

    def mutate(self, rng):
        """Create a mutated offspring at this creature's home position."""
        def _pnz(val, var):
            return max(rng.normal(val, var) if var > 0 else val, FLOAT_MIN_POSITIVE)

        def _pos(val, var):
            return max(rng.normal(val, var) if var > 0 else val, 0.0)

        return Creature(
            pos=self.home_pos.copy(),
            speed=(_pnz(self.speed[0], self.speed[1]), self.speed[1]),
            size=(_pnz(self.size[0], self.size[1]), self.size[1]),
            sense_range=(_pos(self.sense_range_trait[0], self.sense_range_trait[1]),
                         self.sense_range_trait[1]),
            reach=(_pnz(self.reach_trait[0], self.reach_trait[1]), self.reach_trait[1]),
            flee_distance=(_pos(self.flee_distance[0], self.flee_distance[1]),
                           self.flee_distance[1]),
            life_span=(_pnz(self.life_span[0], self.life_span[1]), self.life_span[1]),
            energy=self.energy,
            age=0,
        )

    def grow_older(self):
        """Parent survives to next generation with age + 1."""
        return Creature(
            pos=self.home_pos.copy(),
            speed=self.speed,
            size=self.size,
            sense_range=self.sense_range_trait,
            reach=self.reach_trait,
            flee_distance=self.flee_distance,
            life_span=self.life_span,
            energy=self.energy,
            age=self.age + 1,
        )

    def clone_offspring(self):
        """Identical offspring (for OPT condition). Same traits, age 0."""
        return Creature(
            pos=self.home_pos.copy(),
            speed=self.speed,
            size=self.size,
            sense_range=self.sense_range_trait,
            reach=self.reach_trait,
            flee_distance=self.flee_distance,
            life_span=self.life_span,
            energy=self.energy,
            age=0,
        )


def _dist(a, b):
    """Fast 2D Euclidean distance using Python math (avoids numpy scalar overhead)."""
    dx = float(a[0]) - float(b[0])
    dy = float(a[1]) - float(b[1])
    return math.sqrt(dx * dx + dy * dy)
